packages/gatsby/src/db/index.js

```diff
@@ -0,0 +1,276 @@
+let db
+
+function startDb(saveFile) {
+  return new Promise((resolve, reject) => {
+    const dbOptions = {
+      autoload: true,
+      autoloadCallback: err => {
+        if (err) {
+          reject(err)
+        } else {
+          resolve()
+        }
+      },
+      autosave: true,
+      autosaveInterval: 1000,
+    }
+    db = new loki(saveFile, dbOptions)
+  })
+}
+
+/**
+ * Starts a loki database. If the file already exists, it will be
+ * loaded as the database state. If not, a new database will be
+ * created.
+ *
+ * @param {string} saveFile on disk file that the database will be
+ * saved and loaded from.
+ * @returns {Promise} promise that is resolved once the database and
+ * (optionally) the existing state has been loaded
+ */
+async function start({ saveFile }) {
+  if (!_.isString(saveFile)) {
+    throw new Error(`saveFile must be a path`)
+  }
+  const saveDir = path.dirname(saveFile)
+  await fs.ensureDir(saveDir)
+  await startDb(saveFile)
+}
+
+/**
+ * Returns a reference to the database. If undefined, the db has not been
+ * initalized yet. Call `start()`
+ *
+ * @returns {Object} database, or undefined
+ */
+function getDb() {
+  return db
+}
+
+/**
+ * Deletes all collections.
+ */
+function deleteAllCollections() {
+  if (db) {
+    _.forEach(db.listCollections(), collInfo => {
+      const coll = db.getCollection(collInfo.name)
+      coll.clear({ removeIndices: true })
+      db.removeCollection(collInfo.name)
+    })
+  }
+}
+
+/**
+ * Deletes all collections that are empty
+ */
+function deleteEmptyCollections() {
+  if (db) {
+    _.forEach(db.listCollections(), collInfo => {
+      const coll = db.getCollection(collInfo.name)
+      if (coll.count() === 0) {
+        db.removeCollection(collInfo.name)
+      }
+    })
+  }
+}
+
+/////////////////////////////////////////////////////////////////////
+// Insertions/Updates/Deletions
+/////////////////////////////////////////////////////////////////////
+
+/**
+ * Creates a node in the DB. Will create a collection for the node
+ * type if one hasn't been created yet
+ *
+ * @param {Object} node The ndoe to add. Must have an `id` and
+ * `internal.type`
+ */
+function createNode(node) {
+  invariant(node.internal, `node has no "internal" field`)
+  invariant(node.internal.type, `node has no "internal.type" field`)
+  invariant(node.id, `node has no "id" field`)
+
+  const type = node.internal.type
+
+  let coll = db.getCollection(type)
+  if (!coll) {
+    coll = db.addCollection(type, { unique: [`id`], indices: [`id`] })
+  }
+
+  return coll.insert(node)
+}
+
+/**
+ * Updates a node in the DB
+ *
+ * @param {Object} node The new node information. Will be merged over
+ * the old one (shallow merge)
+ * @param {Object} oldNode The old node to merge the new node
+ * over. Optional. If not supplied, the old node is found by querying
+ * by node.id
+ */
+function updateNode(node, oldNode) {
+  invariant(node.internal, `node has no "internal" field`)
```

Comment:
I like the use of invariant here (and elsewhere!)

Reply:
Yeah, my approach with invariants like this is to be quite liberal in their application, and then perhaps remove them in the future if this code becomes very stable.